From da936173dc5ba5d8a5f5c73010ca844b6a234f72 Mon Sep 17 00:00:00 2001
From: Michael Knopke <knopkem@gmail.com>
Date: Mon, 27 Jul 2026 14:38:30 +0200
Subject: [PATCH 4/9] Fix worker thread detection in parallel mesh processing

WorkerThreadPool group tasks carry no task ID, so get_caller_task_id()
returns INVALID_TASK_ID inside group tasks. Both parallel mesh
processors therefore assumed they weren't running on a pool thread and
dispatched nested group tasks, deadlocking the pool once enough import
workers were blocked in wait_for_group_task_completion().

Use get_thread_index() instead, which correctly identifies pool
threads regardless of task type.
---
 editor/import/3d/resource_importer_scene.cpp | 2 +-
 modules/gltf/gltf_document.cpp               | 2 +-
 2 files changed, 2 insertions(+), 2 deletions(-)

diff --git a/editor/import/3d/resource_importer_scene.cpp b/editor/import/3d/resource_importer_scene.cpp
index bd5f59ee..f7e7f9ca 100644
--- a/editor/import/3d/resource_importer_scene.cpp
+++ b/editor/import/3d/resource_importer_scene.cpp
@@ -2979,7 +2979,7 @@ Node *ResourceImporterScene::_generate_meshes(Node *p_node, const Dictionary &p_
 	Mutex save_mutex;
 	MeshProcessBatchData batch_data = { mesh_tasks.ptrw(), &p_src_lightmap_cache, &r_lightmap_caches, &lightmap_mutex, &save_mutex };
 
-	if (mesh_tasks.size() > 1 && WorkerThreadPool::get_singleton()->get_caller_task_id() == WorkerThreadPool::INVALID_TASK_ID) {
+	if (mesh_tasks.size() > 1 && WorkerThreadPool::get_singleton()->get_thread_index() == -1) {
 		// Not on a worker thread: process meshes in parallel. A worker thread blocking on
 		// wait_for_group_task_completion() can starve the pool, so import threads process inline.
 		WorkerThreadPool::GroupID group = WorkerThreadPool::get_singleton()->add_native_group_task(_process_mesh_worker, &batch_data, mesh_tasks.size(), -1, false, SNAME("SceneMeshProcessing"));
diff --git a/modules/gltf/gltf_document.cpp b/modules/gltf/gltf_document.cpp
index 64af5970..4577b532 100644
--- a/modules/gltf/gltf_document.cpp
+++ b/modules/gltf/gltf_document.cpp
@@ -2057,7 +2057,7 @@ Error GLTFDocument::_parse_meshes(Ref<GLTFState> p_state) {
 	data.errors = &errors;
 	data.material_mutex = &material_mutex;
 
-	if (mesh_count > 1 && WorkerThreadPool::get_singleton()->get_caller_task_id() == WorkerThreadPool::INVALID_TASK_ID) {
+	if (mesh_count > 1 && WorkerThreadPool::get_singleton()->get_thread_index() == -1) {
 		// Not on a worker thread: parse meshes in parallel. A worker thread blocking on
 		// wait_for_group_task_completion() can starve the pool, so import threads parse inline.
 		WorkerThreadPool::GroupID group = WorkerThreadPool::get_singleton()->add_template_group_task(this, &GLTFDocument::_process_single_mesh, &data, mesh_count, -1, false, SNAME("GLTFMeshParsing"));
-- 
2.43.0

